rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py

```diff
@@ -0,0 +1,105 @@
+from typing import Callable, List, Optional
+
+from rotkehlchen.accounting.structures.base import HistoryBaseEntry
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.assets.asset import EvmToken
+from rotkehlchen.chain.ethereum.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.ethereum.decoding.structures import ActionItem
+from rotkehlchen.chain.ethereum.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.chain.ethereum.structures import EthereumTxReceiptLog
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.constants.assets import A_ETH
+from rotkehlchen.constants.misc import ZERO
+from rotkehlchen.types import EthereumTransaction
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+from ..constants import CPT_UNISWAP_V2
+
+# https://www.4byte.directory/api/v1/event-signatures/?hex_signature=0xd78ad95fa46c994b6551d0da85fc275fe613ce37657fb8d5e3d130840159d822  # noqa: E501
+SWAP_SIGNATURE = b'\xd7\x8a\xd9_\xa4l\x99KeQ\xd0\xda\x85\xfc\'_\xe6\x13\xce7e\x7f\xb8\xd5\xe3\xd10\x84\x01Y\xd8"'  # noqa: E501
+
+
+class Uniswapv2Decoder(DecoderInterface):  # lgtm[py/missing-call-to-init]
+
+    def _maybe_decode_v2_swap(  # pylint: disable=no-self-use
+            self,
+            token: Optional[EvmToken],  # pylint: disable=unused-argument
+            tx_log: EthereumTxReceiptLog,
+            transaction: EthereumTransaction,  # pylint: disable=unused-argument
+            decoded_events: List[HistoryBaseEntry],
+            action_items: List[ActionItem],  # pylint: disable=unused-argument
+    ) -> None:
+        """Decode trade for uniswap v2. The approach is to read the events and detect the ones
+        where the user sends and receives any asset. The spend asset is the swap executed and
+        the received is the result of calling the swap method in the uniswap contract.
+        We need to make sure that the events related to the swap are consecutive and for that
+        we make use of the maybe_reshuffle_events.
+        """
+        if tx_log.topics[0] == SWAP_SIGNATURE:
+            # When the router chains multiple swaps in one transaction only the last swap has
+            # the buyer in the topic. In that case we know it is the last swap and the receiver is
+            # the user
+            maybe_buyer = hex_or_bytes_to_address(tx_log.topics[2])
+            out_event = in_event = None
+
+            # amount_in is the amount that enters the pool and amount_out the one
+            # that leaves the pool
+            amount_in_0, amount_in_1 = hex_or_bytes_to_int(tx_log.data[0:32]), hex_or_bytes_to_int(tx_log.data[32:64])  # noqa: E501
```

Comment:
Ahaaaaaaaaaa I think I got it. And if I got it correctly I think that decoding mechanism could be improved.

If I got it correctly, each swapping log event is an event of one particular pool. And this pool can be found in `tx_log.address`. Each pool always swaps 2 tokens. So If we know which pool works with which tokens, we can understand directly from the `tx_log` what assets we are swapping and there is no need in matching events by amount (e.g. we can get rid of calculation of `received_eth`, checks like `event.balance.amount == asset_normalized_value(amount_out, event.asset)`, etc.).

Now about `amount_in_1`,  `amount_in_2`, `amount_out_1` and `amount_out_2`. As mentioned before, all pools have 2 tokens that they work with. So I think it means that if `amount_in_1` and `amount_out_2` are not zeros, then we have a swap of asset A to asset B. And if `amount_in_2` and `amount_out_1` are not zeros, then we have a swap of asset B to asset A.

Am I right, what do you think? If I am, I think this decoder can be greatly simplified

Reply:
What do you mean exactly? Could you write some pseucodecode or explain the changes either here or in discord in more detail?

Reply:
If I got you your idea breaks in `So If we know which pool works with which tokens` because we can't know that without making extra calls in most of the cases and that would be a bad idea during the decoding process (slower and can fail)